libs/go/libp2p_node/acn/utils.go

```diff
@@ -181,3 +200,260 @@ type Pipe interface {
+	if err != nil {
+		return nil, err
+	}
+	var lookupResponse *LookupResponsePerformative = nil
+	var status *StatusPerformative = nil
+	switch pl := response.Performative.(type) {
+	case *LookupResponse:
+		lookupResponse = pl.LookupResponse
+	case *Status:
+		status = pl.Status
+	default:
+		err = errors.New("Unexpected Acn Message")
+		logger.Error().Str("err", err.Error()).Msgf("couldn't deserialize acn lookup response message")
+		return nil, err
+	}
+
+	if status != nil {
+		err = errors.New(
+			"Failed agent lookup response " + status.Body.Code.String() + " : " + strings.Join(
+				status.Body.Msgs,
+				":",
+			),
+		)
+		return nil, err
+	}
+	return lookupResponse.Record, nil
+}
+
+func SendLookupResponse(pipe Pipe, record *AgentRecord) error {
+	lookupResponse := &LookupResponsePerformative{Record: record}
+	response := &AcnMessage{
+		Performative: &LookupResponse{LookupResponse: lookupResponse},
+	}
+	buf, err := proto.Marshal(response)
+	if err != nil {
+		return err
+	}
+	err = pipe.Write(buf)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func SendEnvelopeMessage(pipe Pipe, envelope_bytes []byte, record *AgentRecord) error {
+	acnMsgBytes, err := EncodeAcnEnvelope(envelope_bytes, record)
+	if err != nil {
+		return err
+	}
+	err = pipe.Write(acnMsgBytes)
+	if err != nil {
+		logger.Error().
+			Str("err", err.Error()).
+			Msgf("on pipe write")
+		return err
+	}
+	return nil
+}
+
+func SendAgentRegisterMessage(pipe Pipe, agentRecord *AgentRecord) error {
+	registration := &RegisterPerformative{Record: agentRecord}
+	msg := &AcnMessage{
+		Performative: &Register{Register: registration},
+	}
+	buf, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	err = pipe.Write(buf)
+	if err != nil {
+		return err
+	}
+
+	status, err := ReadAcnStatus(pipe)
+	if err != nil {
+		return err
+	}
+	if status.Code != SUCCESS {
+		return errors.New("Registration failed: " + strings.Join(status.Msgs, ":"))
+	}
+	return nil
+}
+
+func ReadAcnStatus(pipe Pipe) (*StatusBody, error) {
+	buf, err := pipe.Read()
+	if err != nil {
+		logger.Error().
+			Str("err", err.Error()).
+			Msgf("on pipe read")
+		return nil, err
+	}
+
+	response := &AcnMessage{}
+	err = proto.Unmarshal(buf, response)
+	if err != nil {
+		logger.Error().
+			Str("err", err.Error()).
+			Msgf("on acn decode")
+		return nil, err
+	}
+
+	// response is expected to be a Status
+	var status *StatusPerformative
+	switch pl := response.Performative.(type) {
+	case *Status:
+		status = pl.Status
+	default:
+		err = errors.New("Unexpected Acn Message")
+		return nil, err
+	}
+
+	return status.Body, nil
+}
+
+func ReadEnvelopeMessage(pipe Pipe) (*AeaEnvelopePerformative, error) {
+	buf, err := pipe.Read()
+	if err != nil {
+		return nil, err
+	}
+	messageType, envelope, _, acnErr := DecodeAcnMessage(buf)
+
+	if acnErr != nil {
+		err = SendAcnError(
+			pipe,
+			acnErr.Error(),
+			acnErr.ErrorCode,
+		)
+		ignore(err)
+		return nil, acnErr.Err
```

Comment:
no error message?

Reply:
this is ok, errors are logged in function called above, so we just propagate error raised on decode acn message